students/ThanhNhan/lesson07/test_html_render.py

```diff
@@ -0,0 +1,459 @@
+    #assert False
+
+def test_listitem_element():
+
+    e = Li("first", id="TheList", style="line-height:200%")
+    e.append("second")
+    file_contents = render_result(e).strip()
+
+    print(file_contents)  # so we can see it if the test fails
+    assert file_contents.startswith("<li ")
+    assert file_contents.endswith("</li>")
+
+    assert 'style="line-height:200%"' in file_contents
+    assert 'id="TheList"' in file_contents
+
+    # # just to be sure -- there should be a closing bracket to the opening tag
+    assert file_contents[:-1].index(">") > file_contents.index('id="TheList"')
+    assert file_contents[:file_contents.index(">")].count(" ") == 2
+    #assert False
+
+def test_header():
+
+    e = H(2, "Python Header Level 2")
+    file_contents = render_result(e).strip()
+    print(file_contents)  # so we can see it if the test fails
+    assert file_contents.startswith("<h2>")
+    assert file_contents.endswith("</h2>")
+
+    e = H(3, "Python Header Level 3")
+    file_contents = render_result(e).strip()
+    print(file_contents)  # so we can see it if the test fails
+    assert file_contents.startswith("<h3>")
+    assert file_contents.endswith("</h3>")
+
+
+
+########
+# Step 8
+########
+
+def test_meta():
+    """a simple horizontal rule with no attributes"""
+    meta = Meta(charset="UTF-8")
+    file_contents = render_result(meta)
+    print(file_contents)
+    assert file_contents == '<meta charset="UTF-8" />\n'
+
+
+# #####################
+# # indentation testing
+# #  Uncomment for Step 9 -- adding indentation
+# #####################
+
+
+def test_indent():
+    """
+    Tests that the indentation gets passed through to the renderer
+    """
+    html = Html("some content")
+    file_contents = render_result(html, ind="   ").rstrip()  #remove the end newline
+
+    print(file_contents)
+    lines = file_contents.split("\n")
+    assert lines[0].startswith("   <")
+    print(repr(lines[-1]))
+    assert lines[-1].startswith("   <")
+
+
+def test_indent_contents():
+    """
+    The contents in a element should be indented more than the tag
+    by the amount in the indent class attribute
+    """
+    html = Element("some content")
+    file_contents = render_result(html, ind="")
+    print(file_contents)
+    lines = file_contents.split("\n")
+    print(lines)
+    print(lines[1].startswith(Element.indent))
+    assert lines[1].startswith(Element.indent)
+
+
+def test_multiple_indent():   
+#     """
+#     make sure multiple levels get indented fully
+#     """
+    body = Body()
+    body.append(P("some text"))
+    html = Html(body)
+
+    file_contents = render_result(html)
+
+    print(file_contents)
+    lines = file_contents.split("\n")
+    
+    
+    for i in range(3):  # this needed to be adapted to the <DOCTYPE> tag
+        assert lines[i + 1].startswith(i * Element.indent + "<")
+
+
+    assert lines[4].startswith(3 * Element.indent + "some")   
+
+# def test_element_indent1():
```

Comment:
you should uncomment this test -- the content got added to the previous test, but it's better to have separate tests, so if something goes wrong, you know what it it.